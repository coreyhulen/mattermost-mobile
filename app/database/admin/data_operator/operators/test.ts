// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Q} from '@nozbe/watermelondb';

import {MM_TABLES} from '@constants/database';
import App from '@typings/database/app';

import DatabaseManager, {DatabaseType} from '../../database_manager';
import DataOperator from '../index';
import {IsolatedEntities, OperationType} from '../types';
import {
    operateAppRecord,
    operateCustomEmojiRecord,
    operateGlobalRecord,
    operateRoleRecord,
    operateServersRecord,
    operateSystemRecord,
    operateTermsOfServiceRecord,
} from './index';

jest.mock('../../database_manager');

const {APP} = MM_TABLES.DEFAULT;

describe('*** DataOperator: Operators tests ***', () => {
    const createConnection = async (setActive = false) => {
        const dbName = 'server_schema_connection';
        const serverUrl = 'https://appv2.mattermost.com';
        const database = await DatabaseManager.createDatabaseConnection({
            shouldAddToDefaultDatabase: true,
            databaseConnection: {
                actionsEnabled: true,
                dbName,
                dbType: DatabaseType.SERVER,
                serverUrl,
            },
        });

        if (setActive) {
            await DatabaseManager.setActiveServerDatabase({
                displayName: dbName,
                serverUrl,
            });
        }

        return database;
    };

    it('=> operateAppRecord: should return an array of type App ', async () => {
        expect.assertions(3);

        const database = await DatabaseManager.getDefaultDatabase();
        expect(database).toBeTruthy();

        const preparedRecords = await operateAppRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {
                buildNumber: 'build-7',
                createdAt: 1,
                id: 'id-18',
                versionNumber: 'v-1',
            },
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('App');
    });

    it('=> operateGlobalRecord: should return an array of type Global', async () => {
        expect.assertions(3);

        const database = await DatabaseManager.getDefaultDatabase();
        expect(database).toBeTruthy();

        const preparedRecords = await operateGlobalRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {id: 'g-1', name: 'g-n1', value: 'g-v1'},
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('Global');
    });

    it('=> operateServersRecord: should return an array of type Servers', async () => {
        expect.assertions(3);

        const database = await DatabaseManager.getDefaultDatabase();
        expect(database).toBeTruthy();

        const preparedRecords = await operateServersRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {
                dbPath: 'mm-server',
                displayName: 's-displayName',
                id: 's-1',
                mentionCount: 1,
                unreadCount: 0,
                url: 'https://community.mattermost.com',
            },
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('Servers');
    });

    it('=> operateCustomEmojiRecord: should return an array of type CustomEmoji', async () => {
        expect.assertions(3);

        const database = await createConnection();
        expect(database).toBeTruthy();

        const preparedRecords = await operateCustomEmojiRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {id: 'emo-1', name: 'emoji'},
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('CustomEmoji');
    });

    it('=> operateRoleRecord: should return an array of type Role ', async () => {
        expect.assertions(3);

        const database = await createConnection();
        expect(database).toBeTruthy();

        const preparedRecords = await operateRoleRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {id: 'role-1', name: 'role-name-1', permissions: []},
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('Role');
    });

    it('=> operateSystemRecord: should return an array of type System', async () => {
        expect.assertions(3);

        const database = await createConnection();
        expect(database).toBeTruthy();

        const preparedRecords = await operateSystemRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {id: 'system-1', name: 'system-name-1', value: 'system'},
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch('System');
    });

    it('=> operateTermsOfServiceRecord: should return an array of type TermsOfService ', async () => {
        expect.assertions(3);

        const database = await createConnection();
        expect(database).toBeTruthy();

        const preparedRecords = await operateTermsOfServiceRecord({
            database: database!,
            optType: OperationType.CREATE,
            value: {id: 'system-1', acceptedAt: 1},
        });

        expect(preparedRecords).toBeTruthy();
        expect(preparedRecords!.collection.modelClass.name).toMatch(
            'TermsOfService',
        );
    });

    it('=> should create a record in the App table in the default database', async () => {
        expect.assertions(2);

        // Creates a record in the App table
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-1',
                    createdAt: 1,
                    id: 'id-1',
                    versionNumber: 'version-1',
                },
            ],
        });

        // Do a query and find out if the value has been registered in the App table of the default database
        const connection = await DatabaseManager.getDefaultDatabase();
        expect(connection).toBeTruthy();

        const records = (await connection!.collections.
            get(APP).
            query(Q.where('id', 'id-1')).
            fetch()) as App[];

        // We should expect to have a record returned as dictated by our query
        expect(records.length).toBe(1);
    });

    it('=> should create several records in the App table in the default database', async () => {
        expect.assertions(2);

        // Creates a record in the App table
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-10',
                    createdAt: 1,
                    id: 'id-10',
                    versionNumber: 'version-10',
                },
                {
                    buildNumber: 'build-11',
                    createdAt: 1,
                    id: 'id-11',
                    versionNumber: 'version-11',
                },
                {
                    buildNumber: 'build-12',
                    createdAt: 1,
                    id: 'id-12',
                    versionNumber: 'version-12',
                },
                {
                    buildNumber: 'build-13',
                    createdAt: 1,
                    id: 'id-13',
                    versionNumber: 'version-13',
                },
            ],
        });

        // Do a query and find out if the value has been registered in the App table of the default database
        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        const records = (await defaultDB!.collections.
            get(APP).
            query(Q.where('id', Q.oneOf(['id-10', 'id-11', 'id-12', 'id-13']))).
            fetch()) as App[];

        // We should expect to have 4 records created
        expect(records.length).toBe(4);
    });

    it('=> should update a record in the App table in the default database', async () => {
        expect.assertions(3);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        // Update record having id 'id-1'
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.UPDATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-13-13',
                    createdAt: 1,
                    id: 'id-1',
                    versionNumber: 'version-1',
                },
            ],
        });

        const records = (await defaultDB!.collections.
            get(APP).
            query(Q.where('id', 'id-1')).
            fetch()) as App[];
        expect(records.length).toBeGreaterThan(0);

        // Verify if the buildNumber for this record has been updated
        expect(records[0].buildNumber).toMatch('build-13-13');
    });

    it('=> should update several records in the App table in the default database', async () => {
        expect.assertions(4);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        // Update records having id 'id-10' and 'id-11'
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.UPDATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-10x',
                    createdAt: 1,
                    id: 'id-10',
                    versionNumber: 'version-10',
                },
                {
                    buildNumber: 'build-11y',
                    createdAt: 1,
                    id: 'id-11',
                    versionNumber: 'version-11',
                },
            ],
        });

        const records = (await defaultDB!.collections.
            get(APP).
            query(Q.where('id', Q.oneOf(['id-10', 'id-11']))).
            fetch()) as App[];
        expect(records.length).toBe(2);

        // Verify if the buildNumber for those two record has been updated
        expect(records[0].buildNumber).toMatch('build-10x');
        expect(records[1].buildNumber).toMatch('build-11y');
    });

    it('=> [EDGE CASE] should UPDATE instead of CREATE record for existing id', async () => {
        expect.assertions(3);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        // id-10 and id-11 exist but yet the optType is CREATE.  The operator should then prepareUpdate the records instead of prepareCreate
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-10x',
                    createdAt: 1,
                    id: 'id-10',
                    versionNumber: 'version-10',
                },
                {
                    buildNumber: 'build-11x',
                    createdAt: 1,
                    id: 'id-11',
                    versionNumber: 'version-11',
                },
            ],
        });

        const records = (await defaultDB!.collections.
            get(APP).
            query(Q.where('id', Q.oneOf(['id-10', 'id-11']))).
            fetch()) as App[];

        // Verify if the buildNumber for those two record has been updated
        expect(records[0].buildNumber).toMatch('build-10x');
        expect(records[1].buildNumber).toMatch('build-11x');
    });

    it('=> [EDGE CASE] should CREATE instead of UPDATE record for non-existing id', async () => {
        expect.assertions(3);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        // id-15 and id-16 do not exist but yet the optType is UPDATE.  The operator should then prepareCreate the records instead of prepareUpdate
        await DataOperator.handleIsolatedEntity({
            optType: OperationType.UPDATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-10x',
                    createdAt: 1,
                    id: 'id-15',
                    versionNumber: 'version-10',
                },
                {
                    buildNumber: 'build-11x',
                    createdAt: 1,
                    id: 'id-16',
                    versionNumber: 'version-11',
                },
            ],
        });

        const records = (await defaultDB!.collections.
            get(APP).
            query(Q.where('id', Q.oneOf(['id-15', 'id-16']))).
            fetch()) as App[];

        // Verify if the buildNumber for those two record has been created
        expect(records[0].buildNumber).toMatch('build-10x');
        expect(records[1].buildNumber).toMatch('build-11x');
    });

    // TODO :  new operators
});
