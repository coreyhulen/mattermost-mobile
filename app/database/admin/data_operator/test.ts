// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {MM_TABLES} from '@constants/database';

import DatabaseManager, {DatabaseType} from '../database_manager';
import DataOperator from './index';
import {
    operateAppRecord,
    operateCustomEmojiRecord,
    operateDraftRecord,
    operateGlobalRecord,
    operateRoleRecord,
    operateServersRecord,
    operateSystemRecord,
    operateTermsOfServiceRecord,
} from './operators';
import {OperationType, IsolatedEntities} from './types';

jest.mock('../database_manager');

const {DRAFT} = MM_TABLES.SERVER;

describe('*** DataOperator: Handlers tests ***', () => {
    const createConnection = async (setActive = false) => {
        const dbName = 'server_schema_connection';
        const serverUrl = 'https://appv2.mattermost.com';
        const database = await DatabaseManager.createDatabaseConnection({
            shouldAddToDefaultDatabase: true,
            databaseConnection: {
                actionsEnabled: true,
                dbName,
                dbType: DatabaseType.SERVER,
                serverUrl,
            },
        });

        if (setActive) {
            await DatabaseManager.setActiveServerDatabase({
                displayName: dbName,
                serverUrl,
            });
        }

        return database;
    };

    it('=> HandleApp: should write to APP entity', async () => {
        expect.assertions(2);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');

        const data = {
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.APP,
            values: [
                {
                    buildNumber: 'build-10x',
                    createdAt: 1,
                    id: 'id-21',
                    versionNumber: 'version-10',
                },
                {
                    buildNumber: 'build-11y',
                    createdAt: 1,
                    id: 'id-22',
                    versionNumber: 'version-11',
                },
            ],
        };

        await DataOperator.handleIsolatedEntity(data);

        expect(spyOnHandleBase).toHaveBeenCalledWith({
            ...data,
            recordOperator: operateAppRecord,
        });
    });

    it('=> HandleGlobal: should write to GLOBAL entity', async () => {
        expect.assertions(2);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');

        const data = {
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.GLOBAL,
            values: [
                {id: 'global-1-id', name: 'global-1-name', value: 'global-1-value'},
            ],
        };

        await DataOperator.handleIsolatedEntity(data);

        expect(spyOnHandleBase).toHaveBeenCalledWith({
            ...data,
            recordOperator: operateGlobalRecord,
        });
    });

    it('=> HandleServers: should write to SERVERS entity', async () => {
        expect.assertions(2);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');

        const data = {
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.SERVERS,
            values: [
                {
                    dbPath: 'server.db',
                    displayName: 'community',
                    id: 'server-id-1',
                    mentionCount: 0,
                    unreadCount: 0,
                    url: 'https://community.mattermost.com',
                },
            ],
        };

        await DataOperator.handleIsolatedEntity(data);

        expect(spyOnHandleBase).toHaveBeenCalledWith({
            ...data,
            recordOperator: operateServersRecord,
        });
    });

    it('=> HandleCustomEmoji: should write to CUSTOM_EMOJI entity', async () => {
        expect.assertions(2);

        const database = await createConnection(true);
        expect(database).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');

        const data = {
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.CUSTOM_EMOJI,
            values: [
                {
                    id: 'custom-emoji-id-1',
                    name: 'custom-emoji-1',
                },
            ],
        };

        await DataOperator.handleIsolatedEntity(data);

        expect(spyOnHandleBase).toHaveBeenCalledWith({
            ...data,
            recordOperator: operateCustomEmojiRecord,
        });
    });

    it('=> HandleRole: should write to ROLE entity', async () => {
        expect.assertions(2);

        const database = await createConnection(true);
        expect(database).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');

        const data = {
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.ROLE,
            values: [
                {
                    id: 'custom-emoji-id-1',
                    name: 'custom-emoji-1',
                    permissions: ['custom-emoji-1'],
                },
            ],
        };

        await DataOperator.handleIsolatedEntity(data);

        expect(spyOnHandleBase).toHaveBeenCalledWith({
            ...data,
            recordOperator: operateRoleRecord,
        });
    });

    it('=> HandleSystem: should write to SYSTEM entity', async () => {
        expect.assertions(2);

        const database = await createConnection(true);
        expect(database).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');

        const data = {
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.SYSTEM,
            values: [{id: 'system-id-1', name: 'system-1', value: 'system-1'}],
        };

        await DataOperator.handleIsolatedEntity(data);

        expect(spyOnHandleBase).toHaveBeenCalledWith({
            ...data,
            recordOperator: operateSystemRecord,
        });
    });

    it('=> HandleTermsOfService: should write to TERMS_OF_SERVICE entity', async () => {
        expect.assertions(2);

        const database = await createConnection(true);
        expect(database).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');

        const data = {
            optType: OperationType.CREATE,
            tableName: IsolatedEntities.TERMS_OF_SERVICE,
            values: [{id: 'tos-1', acceptedAt: 1}],
        };

        await DataOperator.handleIsolatedEntity(data);

        expect(spyOnHandleBase).toHaveBeenCalledWith({
            ...data,
            recordOperator: operateTermsOfServiceRecord,
        });
    });

    it('=> No table name: should not call handleBase if tableName is invalid', async () => {
        expect.assertions(2);

        const defaultDB = await DatabaseManager.getDefaultDatabase();
        expect(defaultDB).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');

        const data = {
            optType: OperationType.CREATE,
            tableName: 'INVALID_TABLE_NAME',
            values: [{id: 'tos-1', acceptedAt: 1}],
        };

        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        await DataOperator.handleIsolatedEntity(data);

        expect(spyOnHandleBase).toHaveBeenCalledTimes(0);
    });

    it('=> handleReactions: should write to both Reactions and CustomEmoji entities', async () => {
        expect.assertions(3);

        const database = await createConnection(true);
        expect(database).toBeTruthy();

        const spyOnPrepareBase = jest.spyOn(DataOperator as any, 'prepareBase');
        const spyOnBatchOperation = jest.spyOn(DataOperator as any, 'batchOperations');

        await DataOperator.handleReactions({
            reactions: [
                {
                    create_at: 1608263728086,
                    delete_at: 0,
                    emoji_name: 'p4p1',
                    post_id: '4r9jmr7eqt8dxq3f9woypzurry',
                    update_at: 1608263728077,
                    user_id: 'ooumoqgq3bfiijzwbn8badznwc',
                },
            ],
            prepareRowsOnly: false,
        });

        // Called twice:  Once for Reaction record and once for CustomEmoji record
        expect(spyOnPrepareBase).toHaveBeenCalledTimes(2);

        // Only one batch operation for both entities
        expect(spyOnBatchOperation).toHaveBeenCalledTimes(1);
    });

    it('=> handleDraft: should write to the Draft entity', async () => {
        expect.assertions(2);

        const database = await createConnection(true);
        expect(database).toBeTruthy();

        const spyOnHandleBase = jest.spyOn(DataOperator as any, 'handleBase');
        const data = [
            {
                channel_id: '4r9jmr7eqt8dxq3f9woypzurrychannelid',
                files: [{
                    user_id: 'user_id',
                    post_id: 'post_id',
                    create_at: 123,
                    update_at: 456,
                    delete_at: 789,
                    name: 'an_image',
                    extension: 'jpg',
                    size: 10,
                    mime_type: 'image',
                    width: 10,
                    height: 10,
                    has_preview_image: false,
                    clientId: 'clientId',
                }],
                message: 'test draft message for post',
                root_id: '',
            },
        ];
        await DataOperator.handleDraft(data);

        // Only one batch operation for both entities
        expect(spyOnHandleBase).toHaveBeenCalledWith({
            optType: OperationType.CREATE,
            tableName: DRAFT,
            values: data,
            recordOperator: operateDraftRecord,
        });
    });

    it('=> handleFiles: should write to File entity', async () => {
        expect.assertions(3);

        const database = await createConnection(true);
        expect(database).toBeTruthy();

        const spyOnPrepareBase = jest.spyOn(DataOperator as any, 'prepareBase');
        const spyOnBatchOperation = jest.spyOn(DataOperator as any, 'batchOperations');

        await DataOperator.handleFiles({
            files: [{
                user_id: 'user_id',
                post_id: 'post_id',
                create_at: 12345,
                update_at: 456,
                delete_at: 789,
                name: 'an_image',
                extension: 'jpg',
                size: 10,
                mime_type: 'image',
                width: 10,
                height: 10,
                has_preview_image: false,
            }],
            prepareRowsOnly: false,
        });

        expect(spyOnPrepareBase).toHaveBeenCalledTimes(1);
        expect(spyOnBatchOperation).toHaveBeenCalledTimes(1);
    });

    it('=> handlePostMetadata: should write to PostMetadata entity', async () => {
        expect.assertions(3);

        const database = await createConnection(true);
        expect(database).toBeTruthy();

        const spyOnPrepareBase = jest.spyOn(DataOperator as any, 'prepareBase');
        const spyOnBatchOperation = jest.spyOn(DataOperator as any, 'batchOperations');

        const data = {
            images: {
                'https://community-release.mattermost.com/api/v4/image?url=https%3A%2F%2Favatars1.githubusercontent.com%2Fu%2F6913320%3Fs%3D400%26v%3D4': {
                    width: 400,
                    height: 400,
                    format: 'png',
                    frame_count: 0,
                },

            },
            embeds: [
                {
                    type: 'opengraph',
                    url: 'https://github.com/mickmister/mattermost-plugin-default-theme',
                    data: {
                        type: 'object',
                        url: 'https://github.com/mickmister/mattermost-plugin-default-theme',
                        title: 'mickmister/mattermost-plugin-default-theme',
                        description: 'Contribute to mickmister/mattermost-plugin-default-theme development by creating an account on GitHub.',
                        determiner: '',
                        site_name: 'GitHub',
                        locale: '',
                        locales_alternate: null,
                        images: [
                            {
                                url: '',
                                secure_url: 'https://community-release.mattermost.com/api/v4/image?url=https%3A%2F%2Favatars1.githubusercontent.com%2Fu%2F6913320%3Fs%3D400%26v%3D4',
                                type: '',
                                width: 0,
                                height: 0,
                            },
                        ],
                        audios: null,
                        videos: null,
                    },
                },
            ],
        };

        await DataOperator.handlePostMetadata({
            embeds: [{
                embed: data.embeds, postId: 'post-1',
            }],
            images: [{
                images: data.images, postId: 'post-1',
            }],
            prepareRowsOnly: false});

        expect(spyOnPrepareBase).toHaveBeenCalledTimes(1);
        expect(spyOnBatchOperation).toHaveBeenCalledTimes(1);
    });

    // TODO : test utils functions (  sanitizeReactions, createPostsChain, sanitizePosts)
});
