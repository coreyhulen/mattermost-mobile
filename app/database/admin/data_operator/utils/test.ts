// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {mockedPosts} from './mock';
import {createPostsChain, sanitizePosts} from './utils';

describe('DataOperator: Utils tests', () => {
    it('sanitizePosts: should filter between ordered and unordered posts', () => {
        const {orderedPosts, unOrderedPosts} = sanitizePosts({posts: Object.values(mockedPosts.posts), orders: mockedPosts.order});
        expect(orderedPosts.length).toBe(4);
        expect(unOrderedPosts.length).toBe(2);
    });

    it('createPostsChain: should link posts amongst each other based on order array', () => {
        const previousPostId = 'prev_xxyuoxmehne';
        const chainedOfPosts = createPostsChain({orders: mockedPosts.order, rawPosts: Object.values(mockedPosts.posts), previousPostId});

        // eslint-disable-next-line max-nested-callbacks
        const post1 = chainedOfPosts.find((post) => {
            return post.id === '8swgtrrdiff89jnsiwiip3y1eoe';
        });
        expect(post1).toBeTruthy();
        expect(post1!.prev_post_id).toBe(previousPostId);

        // eslint-disable-next-line max-nested-callbacks
        const post2 = chainedOfPosts.find((post) => {
            return post.id === '8fcnk3p1jt8mmkaprgajoxz115a';
        });
        expect(post2).toBeTruthy();
        expect(post2!.prev_post_id).toBe('8swgtrrdiff89jnsiwiip3y1eoe');

        // eslint-disable-next-line max-nested-callbacks
        const post3 = chainedOfPosts.find((post) => {
            return post.id === '3y3w3a6gkbg73bnj3xund9o5ic';
        });
        expect(post3).toBeTruthy();
        expect(post3!.prev_post_id).toBe('8fcnk3p1jt8mmkaprgajoxz115a');

        // eslint-disable-next-line max-nested-callbacks
        const post4 = chainedOfPosts.find((post) => {
            return post.id === '4btbnmticjgw7ewd3qopmpiwqw';
        });
        expect(post4).toBeTruthy();
        expect(post4!.prev_post_id).toBe('3y3w3a6gkbg73bnj3xund9o5ic');
    });
});
