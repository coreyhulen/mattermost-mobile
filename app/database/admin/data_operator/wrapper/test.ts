// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import DatabaseManager from '@database/admin/database_manager';
import {createDataOperator} from '@database/admin/data_operator/wrapper';
import DatabaseConnectionException from '@database/admin/exceptions/database_connection_exception';
import {DatabaseType} from '@typings/database/enums';

jest.mock('@database/admin/database_manager');

describe('*** DataOperator Wrapper ***', () => {
    it('=> wrapper should return an instance of DataOperator ', async () => {
        expect.assertions(1);

        const serverUrl = 'https://wrapper.mattermost.com';

        // first we create the connection and save it into default database
        await DatabaseManager.createDatabaseConnection({
            configs: {
                actionsEnabled: true,
                dbName: 'community mattermost',
                dbType: DatabaseType.SERVER,
                serverUrl,
            },
            shouldAddToDefaultDatabase: true,
        });

        const dataOperator = await createDataOperator(serverUrl);

        expect(dataOperator).toBeTruthy();
    });

    it('=> wrapper should throw an error due to invalid server url', async () => {
        expect.assertions(2);

        const serverUrl = 'https://wrapper.mattermost.com';

        // first we create the connection and save it into default database
        await DatabaseManager.createDatabaseConnection({
            configs: {
                actionsEnabled: true,
                dbName: 'test_database',
                dbType: DatabaseType.SERVER,
                serverUrl,
            },
            shouldAddToDefaultDatabase: true,
        });

        await expect(createDataOperator('https://error.com')).rejects.toThrow(
            'No database has been registered with this url: https://error.com',
        );

        await expect(createDataOperator('https://error.com')).rejects.toThrow(DatabaseConnectionException);
    });
});
