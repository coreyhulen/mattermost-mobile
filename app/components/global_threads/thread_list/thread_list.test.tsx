// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {FlatList} from 'react-native';
import * as redux from 'react-redux';
import {shallow} from 'enzyme';

import initialState from '@store/initial_state';
import {intl} from 'test/intl-test-helper';

import ThreadList from './thread_list';

jest.mock('app/components/compass_icon', () => 'Icon');
jest.spyOn(redux, 'useSelector').mockImplementation((callback) => callback(initialState));
jest.spyOn(React, 'useRef').mockReturnValue({
    current: {},
});

describe('Global Thread List', () => {
    const testID = 'thread_list';

    const markAllAsRead = jest.fn();
    const viewAllThreads = jest.fn();
    const viewUnreadThreads = jest.fn();

    const baseProps = {
        haveUnreads: true,
        intl,
        isLoading: false,
        listRef: React.useRef<FlatList>(null),
        loadMoreThreads: jest.fn(),
        markAllAsRead,
        testID,
        threadIds: ['thread1'],
        viewingUnreads: true,
        viewAllThreads,
        viewUnreadThreads,
    };

    const wrapper = shallow(
        <ThreadList
            {...baseProps}
        />,
    );

    test('Should render threads with functional tabs & mark all as read button', () => {
        expect(wrapper.getElement()).toMatchSnapshot();

        const allThreadsTab = wrapper.find({testID: `${testID}.all_threads`}).at(0);
        expect(allThreadsTab.exists()).toBeTruthy();
        allThreadsTab.simulate('press');
        expect(viewAllThreads).toHaveBeenCalled();

        const unreadThreadsTab = wrapper.find({testID: `${testID}.unread_threads`}).at(0);
        expect(unreadThreadsTab.exists()).toBeTruthy();
        unreadThreadsTab.simulate('press');
        expect(viewUnreadThreads).toHaveBeenCalled();

        expect(wrapper.find({testID: `${testID}.unreads_dot`}).exists()).toBeTruthy();

        const markAllAsReadButton = wrapper.find({testID: `${testID}.mark_all_read`}).at(0);
        expect(markAllAsReadButton.exists()).toBeTruthy();
        expect(markAllAsReadButton.props().disabled).toBeFalsy();
        markAllAsReadButton.simulate('press');
        expect(markAllAsRead).toHaveBeenCalled();
    });

    test('Should disable mark all as read and hide dot on UNREADS tab when no unread messages are present', () => {
        wrapper.setProps({
            ...baseProps,
            haveUnreads: false,
        });

        expect(wrapper.find({testID: `${testID}.unreads_dot`}).exists()).toBeFalsy();

        const markAllAsReadButton = wrapper.find({testID: `${testID}.mark_all_read`}).at(0);
        expect(markAllAsReadButton.exists()).toBeTruthy();
        expect(markAllAsReadButton.props().disabled).toBeTruthy();
    });
});
