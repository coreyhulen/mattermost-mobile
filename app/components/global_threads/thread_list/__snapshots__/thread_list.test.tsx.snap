// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global Thread List Should render threads with functional tabs & mark all as read button 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 64,
        },
        Object {
          "borderBottomColor": "rgba(61,60,64,0.08)",
          "borderBottomWidth": 1,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "flexGrow": 1,
        }
      }
    >
      <ForwardRef
        onPress={[MockFunction]}
        testID="thread_list.all_threads"
      >
        <View
          style={
            Array [
              Object {
                "marginLeft": 16,
                "paddingHorizontal": 16,
                "paddingVertical": 8,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "alignSelf": "center",
                  "color": "rgba(61,60,64,0.56)",
                  "fontFamily": "Open Sans",
                  "fontSize": 16,
                  "fontWeight": "600",
                  "lineHeight": 24,
                },
                Object {},
              ]
            }
          >
            All Your Threads
          </Text>
        </View>
      </ForwardRef>
      <ForwardRef
        onPress={[MockFunction]}
        testID="thread_list.unread_threads"
      >
        <View
          style={
            Array [
              Object {
                "marginLeft": 16,
                "paddingHorizontal": 16,
                "paddingVertical": 8,
              },
              Object {
                "backgroundColor": "rgba(22,109,224,0.08)",
                "borderRadius": 4,
              },
            ]
          }
        >
          <View>
            <Text
              style={
                Array [
                  Object {
                    "alignSelf": "center",
                    "color": "rgba(61,60,64,0.56)",
                    "fontFamily": "Open Sans",
                    "fontSize": 16,
                    "fontWeight": "600",
                    "lineHeight": 24,
                  },
                  Object {
                    "color": "#166de0",
                  },
                ]
              }
            >
              Unreads
            </Text>
            <View
              style={
                Object {
                  "backgroundColor": "#579eff",
                  "borderRadius": 3,
                  "height": 6,
                  "position": "absolute",
                  "right": -6,
                  "width": 6,
                }
              }
              testID="thread_list.unreads_dot"
            />
          </View>
        </View>
      </ForwardRef>
    </View>
    <View
      style={
        Object {
          "paddingHorizontal": 20,
        }
      }
    >
      <ForwardRef
        disabled={false}
        onPress={[MockFunction]}
        testID="thread_list.mark_all_read"
      >
        <Icon
          name="playlist-check"
          style={
            Object {
              "color": "rgba(61,60,64,0.56)",
              "fontSize": 28,
              "lineHeight": 28,
            }
          }
        />
      </ForwardRef>
    </View>
  </View>
  <FlatList
    ListEmptyComponent={
      <EmptyState
        intl={
          Object {
            "defaultFormats": Object {},
            "defaultLocale": "en",
            "formatDate": [Function],
            "formatHTMLMessage": [Function],
            "formatMessage": [Function],
            "formatNumber": [Function],
            "formatPlural": [Function],
            "formatRelative": [Function],
            "formatTime": [Function],
            "formats": Object {},
            "formatters": Object {
              "getDateTimeFormat": [Function],
              "getMessageFormat": [Function],
              "getNumberFormat": [Function],
              "getPluralFormat": [Function],
              "getRelativeFormat": [Function],
            },
            "locale": "en",
            "messages": Object {},
            "now": [Function],
            "onError": [Function],
            "textComponent": "span",
            "timeZone": null,
          }
        }
        isUnreads={true}
      />
    }
    contentContainerStyle={
      Object {
        "flexGrow": 1,
      }
    }
    data={
      Array [
        "thread1",
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={[Function]}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;
